farm-bot/codebase/contracts/farm-bot.sol

```diff
@@ -2,62 +2,205 @@
+            return _lpAmount * totalSupply / lpTotalBalance;
         }
     }
 
+    function getLpAmount(uint256 _fpAmount) public view returns (uint256) {
+	if (totalSupply == 0) {
+	    return 0;
+	} else {
+	    return _fpAmount * lpTotalBalance / totalSupply;
+	}
+    }
+
     function deposit(uint256 _lpAmount) public {
-        // todo might need a lock on this
-        bool transferSuccess = IERC20(tokenAddress).transferFrom(msg.sender, address(this), _lpAmount);
+        bool transferSuccess = stakingToken.transferFrom(msg.sender, address(this), _lpAmount);
         require(transferSuccess, "Transfer failed, aborting deposit");
 
         uint256 _fpAmount = this.getFpAmount(_lpAmount);
-        fpTotalBalance += _fpAmount;
+	_mint(msg.sender, _fpAmount);
         lpTotalBalance += _lpAmount;
-        fpBalances[msg.sender] += _fpAmount;
         investInFarm();
     }
 
+    function withdrawAll() public {
+	require(balanceOf[msg.sender] > 0, "Cannot withdraw zero balance");
+	uint256 _lpAmount = getLpAmount(balanceOf[msg.sender]);
+	withdraw(_lpAmount);
+    }
+
     function withdraw(uint256 _lpAmount) public {
         // todo might need a lock on this
         uint256 _fpAmount = this.getFpAmount(_lpAmount);
-        require(fpBalances[msg.sender] >= _fpAmount, "Cannot withdraw more than the total balance of the owner");
+        require(balanceOf[msg.sender] >= _fpAmount, "Cannot withdraw more than the total balance of the owner");
 
-        uint256 tokenBalance = IERC20(tokenAddress).balanceOf(address(this));
+        uint256 tokenBalance = stakingToken.balanceOf(address(this));
         if (_lpAmount > tokenBalance) {
-            farm.withdraw(_lpAmount - tokenBalance);
+            stakingRewards.withdraw(_lpAmount - tokenBalance);
         }
 
-        bool transferSuccess = IERC20(tokenAddress).transfer(msg.sender, _lpAmount);
+        bool transferSuccess = stakingToken.transfer(msg.sender, _lpAmount);
         require(transferSuccess, "Transfer failed, aborting withdrawal");
-        fpBalances[msg.sender] -= _fpAmount;
-        fpTotalBalance -= _fpAmount;
+	_burn(msg.sender, _fpAmount);
         lpTotalBalance -= _lpAmount;
     }
 
-    function investInFarm() public {
-        // todo eventually make private. The public function will do more: reap rewards, swap them, obtain LP tokens, and then this.
-        uint256 tokenBalance = IERC20(tokenAddress).balanceOf(address(this));
+    function investInFarm() private {
+        uint256 tokenBalance = stakingToken.balanceOf(address(this));
         require(tokenBalance > 0, "Cannot invest in farm because tokenBalance is 0");
-        IERC20(tokenAddress).approve(address(farm), tokenBalance);
-        farm.stake(tokenBalance);
+        stakingToken.approve(address(stakingRewards), tokenBalance);
+        stakingRewards.stake(tokenBalance);
+    }
+
+    function claimRewards(uint deadline) public ensure(deadline) {
```

Comment:
nit: this does both claiming rewards and reinvesting, could be broken down a bit